library/core/src/primitive_docs.rs

```diff
@@ -1389,6 +1389,25 @@ mod prim_usize {}
 /// work on references as well as they do on owned values! The implementations described here are
 /// meant for generic contexts, where the final type `T` is a type parameter or otherwise not
 /// locally known.
+///
+/// # Safety
+///
+/// For all types, `T: ?Sized`, and for all `t: &T` or `t: &mut T`, unsafe code may assume that
+/// the following properties hold. Rust programmers must assume that, unless explicitly stated
+/// otherwise, any Rust code they did not author themselves may rely on these properties, and that
+/// violating them may cause that code to exhibit undefined behavior.
+///
+/// * `t` is aligned to `align_of_val(t)`
+/// * `t` is dereferenceable for `size_of_val(t)` many bytes
+///
+/// If `t` points at address `a`, being "dereferenceable" for N bytes means that the memory range
+/// `[a, a + N)` is all contained within a single [allocated object].
+///
+/// Note that the precise validity invariants for reference types are a work in progress. In the
+/// future, new guarantees may be added. However, the guarantees documented in this section will
+/// never be removed.
```

Comment:
The wording here does make it sound like above we're describing a validity invariant.

Maybe we should try to word this more like:
```
When values of reference type cross an API boundary, the following invariant must generally be upheld:

[...]

For instance, this means that unsafe code in a safe function may assume these invariants are ensured by the caller.

It is not decided yet whether unsafe code may violate this invariant temporarily on internal data.
```

Reply:
How about:

> When values of reference type cross an API boundary, the following invariant must generally be upheld:
>
> [...]
>
> For instance, this means that unsafe code in a safe function may assume these invariants are ensured of arguments passed by the caller, and it may assume that these invariants are ensured of return values from any safe functions it calls. The inverse is also true: unsafe code must not violate these invariants when passing arguments to safe functions or returning values from safe functions; such violations may result in undefined behavior.
>
> It is not decided yet whether unsafe code may violate these invariants temporarily on internal data. As a consequence, unsafe code which violates these invariants temporarily on internal data may become unsound in future versions of Rust depending on how this question is decided.

Reply:
> unsafe code must not violate these invariants when passing arguments to safe functions or returning values from safe functions; such violations may result in undefined behavior.

I'm slightly uneasy since ultimately what matters is not whether the other function is safe but whether it is known to have non-standard preconditions... for instance, a function could be written taking `&str` and explicitly saying that it may be called even for non-UTF-8 strings. I am not sure if it's worth going into that level of detail though, we're not teaching safety abstractions here, just defining a safety invariant.

Maybe slap in an "in general" somewhere, to indicate that there are potential exceptions.

The last paragraph seems unnecessarily verbose/repetitive to me, but :shrug: I don't care strongly.

Sounds good otherwise!

Reply:
> I'm slightly uneasy since ultimately what matters is not whether the other function is safe but whether it is known to have non-standard preconditions... for instance, a function could be written taking `&str` and explicitly saying that it may be called even for non-UTF-8 strings. I am not sure if it's worth going into that level of detail though, we're not teaching safety abstractions here, just defining a safety invariant.

I actually don't think that's true. Consider the following function:

```rust
/// Forgets `b`.
///
/// Since `b` is forgotten, the caller is allowed to pass an invalidly-constructed `Box`
/// (e.g., `Box::from_raw(8 as *mut usize)`).
fn forget(b: Box<usize>) {
    std::mem::drop(b);
}
```

Would we want to say that it's okay for `unsafe` code to call `forget(unsafe { Box::from_raw(8 as *mut usize) })`? If we do, then we're effectively saying that we trust the safety promises even of code that hasn't opted in to being allowed to provide safety promises.

I think of safe code kind of like being a legal minor - you are free from a lot of responsibilities, but you're also not legally allowed to enter into contracts on your own. To rephrase that, you're not allowed to make legally binding commitments. It would sort of be like a toddler (safe code) saying, "can you give me that kitchen knife? Trust me, I won't do anything dangerous with it." Obviously an adult (unsafe code) shouldn't trust the kid to keep such a promise.

> The last paragraph seems unnecessarily verbose/repetitive to me

Just trying to sow the seeds of conservatism in the Rust world by constantly reminding people to think in terms of guarantees instead of currently-is-the-case's 😛

Reply:
Unsafe code relies on safe code promising more than the basic safety promises all the time. If I implement a data structure in safe code, and then you write unsafe code trusting my data structure to be functionally correct, then the contents of the safe code becomes part of your TCB.

So, safe code making promises like "you may call this with arguments that violate their safety invariant" seems completely fine to me, similar to safe code promising "this function will return true if and only if the element is in the set" (or any other kind of functional promise).

Reply:
LGTM, please update the PR. :)

Reply:
Done!